### PR TITLE
Improve logic for reading dictionaries during row group pruning

### PR DESCRIPTION
**Pull request by raunaqmorarka**
## Description

Improve logic for reading dictionaries during row group pruning in parquet reader

## Non-technical explanation

Improves parquet reader performance in the presence of predicates.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Hive, Delta, Iceberg
* Improve performance of reading parquet files for queries with predicates. ({issue}`14247`)
```

### CODE REVIEW

File: lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java [hunk 1]
```diff
@@ -635,23 +635,14 @@ private ColumnIndexValueConverter()
...
-                    return buffer -> buffer.order(LITTLE_ENDIAN).getFloat(0);
-                case DOUBLE:
-                    return buffer -> buffer.order(LITTLE_ENDIAN).getDouble(0);
-                case FIXED_LEN_BYTE_ARRAY:
-                case BINARY:
-                case INT96:
-                default:
-                    return buffer -> Binary.fromReusedByteBuffer(buffer);
-            }
+            return switch (primitiveType.getPrimitiveTypeName()) {
```
Comment (skrzypo987) [inline lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:638]: nit: I'd squash the two commits with modern switches

**Review by skrzypo987 — COMMENTED** [review]
Skimmed, looks good

File: lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java (hunk 1 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java]: I had to drop those commits as they were causing the error prone plugin to crash due to some bug of it's own

File: lib/trino-parquet/src/main/java/io/trino/parquet/predicate/Predicate.java [hunk 2]
```diff
@@ -21,20 +21,28 @@
...
+import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 
 public interface Predicate
 {
     /**
-     * Should the Parquet Reader process a file section with the specified statistics.
+     * Should the Parquet Reader process a file section with the specified statistics,
+     * and if it should then which columns are candidates for further inspection of more granular statistics
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/predicate/Predicate.java]: updated the comment

File: lib/trino-parquet/src/main/java/io/trino/parquet/predicate/DictionaryDescriptor.java [hunk 3]
```diff
@@ -21,11 +21,13 @@
 public class DictionaryDescriptor
```
Comment (sopel39) [inline lib/trino-parquet/src/main/java/io/trino/parquet/predicate/DictionaryDescriptor.java:21]: > When the nulls count in column statistics for a column is 0

Do you think it's possible for some writer to set it incorrectly?

**Review by sopel39 — APPROVED** [review]
Do you think we should introduce feature flag for optimized reader first?

File: lib/trino-parquet/src/main/java/io/trino/parquet/predicate/DictionaryDescriptor.java (hunk 3 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/predicate/DictionaryDescriptor.java:21]: If column statistics are incorrect then our existing predicate pushdown logic could also give wrong results, so we should be able to rely on it

**Comment by raunaqmorarka** [comment]
> Do you think we should introduce feature flag for optimized reader first?

These are targeted fixes/improvements in existing code and not a rewrite of the reader, so these shouldn't need to be behind a flag

**Pull request merged**